File: Dynamic Programming/house_robber_iii.cpp
// https://leetcode.com/problems/house-robber-iii/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    map< pair<TreeNode*, bool>, int> dp;
    
    int solve(TreeNode* root, bool OK) {
        if(!root)
            return 0;
        
        if(dp.count({root, OK}))
            return dp[{root, OK}];
        
        if(OK) {
            return dp[{root, OK}] = root->val + solve(root->left, false) + solve(root->right, false);
        }
        
        int left = max(solve(root->left, true), solve(root->left, false));
        int right = max(solve(root->right, true), solve(root->right, false));
        return dp[{root, OK}] = left + right;
    }
    
    int rob(TreeNode* root) {
        return max(solve(root, true), solve(root, false));
    }
};

File: Tree/house_robber_iii.go
// https://leetcode.com/problems/house-robber-iii/description/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Pair[T, U any] struct {
    first T
    second U
}

func solve(root *TreeNode, prevRobbed bool, memory map[Pair[*TreeNode, bool]]int) int {
    if root == nil {
        return 0
    }

    pair := Pair[*TreeNode, bool]{
        first: root,
        second: prevRobbed,
    }

    if val, ok := memory[pair]; ok {
        return val
    }

    if prevRobbed {
        memory[pair] = solve(root.Left, false, memory) + solve(root.Right, false, memory)
        return memory[pair]
    }

    robbing := root.Val + solve(root.Left, true, memory) + solve(root.Right, true, memory)
    noRobbing := solve(root.Left, false, memory) + solve(root.Right, false, memory)

    memory[pair] = max(robbing, noRobbing)

    return memory[pair]
}

func rob(root *TreeNode) int {
    memory := map[Pair[*TreeNode, bool]]int{}

    return max(solve(root, false, memory), solve(root, true, memory)) 
}
